node/network/availability-distribution/src/lib.rs

```diff
@@ -0,0 +1,1856 @@
+	};
+
+	// create proofs for each erasure chunk
+	let branches = branches(chunks.as_ref());
+
+	let erasure_chunks = branches
+		.enumerate()
+		.map(|(index, (proof, chunk))| ErasureChunk {
+			chunk: chunk.to_vec(),
+			index: index as _,
+			proof,
+		})
+		.collect::<Vec<ErasureChunk>>();
+
+	Ok(erasure_chunks)
+}
+
+/// Handle the changes necessary when our view changes.
+async fn handle_our_view_change<Context>(
+	ctx: &mut Context,
+	keystore: KeyStorePtr,
+	state: &mut ProtocolState,
+	view: View,
+) -> Result<()>
+where
+	Context: SubsystemContext<Message = AvailabilityDistributionMessage>,
+{
+	let old_view = std::mem::replace(&mut (state.view), view);
+
+	// needed due to borrow rules
+	let view = state.view.clone();
+	let added = view.difference(&old_view).collect::<Vec<&'_ Hash>>();
+
+	// add all the relay parents and fill the cache
+	for added in added.clone() {
+		let validators = query_validators(ctx, added.clone()).await?;
+		let validator_index = obtain_our_validator_index(
+			&validators,
+			keystore.clone(),
+		);
+		state.add_relay_parent(ctx, added.clone(), validators, validator_index).await?;
+	}
+
+	// handle all candidates
+	for (candidate_hash, _receipt) in state.cached_live_candidates_unioned(added) {
+		let per_candidate = state
+			.per_candidate
+			.entry(candidate_hash.clone())
+			.or_default();
+
+		// assure the node has the validator role
+		let validator_index = if let Some(validator_index) = per_candidate.validator_index {
+			validator_index
+		} else {
+			continue;
+		};
+
+		// pull the proof of validity
+		let available_data = if let Some(available_data) =
+			query_available_data(ctx, candidate_hash.clone()).await?
+		{
+			available_data
+		} else {
+			continue;
+		};
+
+
+		let erasure_chunks = if let Ok(erasure_chunks) = derive_erasure_chunks_with_proofs(
+			per_candidate.validators.len(),
+			validator_index,
+			&available_data,
+		) {
+			erasure_chunks
+		} else {
+			return Ok(());
+		};
+
+		if let Some(erasure_chunk) = erasure_chunks.get(validator_index as usize) {
+			match store_chunk(ctx, candidate_hash, validator_index, erasure_chunk.clone()).await {
+				Err(_e) => warn!(target: TARGET, "Failed to send store message to overseer"),
+				Ok(Err(())) => warn!(target: TARGET, "Failed to store our own erasure chunk"),
+				Ok(Ok(())) => {}
+			}
+		} else {
+			warn!(
+				target: TARGET,
+				"Our validation index is out of bounds, no associated message"
+			);
+			return Ok(());
+		}
+
+		// obtain interested peers in the candidate hash
+		let peers: Vec<PeerId> = state
+			.peer_views
+			.clone()
+			.into_iter()
+			.filter(|(_peer, view)| {
+				// collect all direct interests of a peer w/o ancestors
+				state
+					.cached_live_candidates_unioned(view.0.iter())
+					.contains_key(&candidate_hash)
+			})
+			.map(|(peer, _view)| peer.clone())
+			.collect();
+
+		// distribute all erasure messages to interested peers
+		for erasure_chunk in erasure_chunks {
```

Comment:
sending a message requires `&mut ctx` which can't be shared due to Rust's ownership rules

Reply:
Good point, but `OverseerSubsystemContext<M>` has a `mpsc::Sender` which is intentionally cheap to clone.

It would be very straightforward to add a method to `Context` like `fn sender(&self) -> mpsc::Sender<ToOverseer>`, but that might be too much. It would be less straightforward, but a better abstraction, to add a message like `fn message_sender<Message: Into<AllMessages>>(&self) -> mpsc::Sender<Message>`, and then spawn a little job which forwards everything from the returned sender to the overseer, constructing a `ToOverseer` message from it on the way.

In both cases, we could add the method as a default method on the trait, which keeps things simple for the implementer.

Reply:
I think this is equiv to the other topic, the overseer processes the messages in sequence as part of a `mpsc::channel()`, so panellizing does not provide many advantages, but a few risks: It spikes the number of elements being in the channel, potentially exhausting the channels capacity. It also would only cause a very short term concurrent generation of messages, which I doubt will be any faster given the overhead of launching the task on a thread in the threadpool.

Reply:
> Good point, but `OverseerSubsystemContext<M>` has a `mpsc::Sender` which is intentionally cheap to clone.
> 
> It would be very straightforward to add a method to `Context` like `fn sender(&self) -> mpsc::Sender<ToOverseer>`, but that might be too much. It would be less straightforward, but a better abstraction, to add a message like `fn message_sender<Message: Into<AllMessages>>(&self) -> mpsc::Sender<Message>`, and then spawn a little job which forwards everything from the returned sender to the overseer, constructing a `ToOverseer` message from it on the way.
> 
> In both cases, we could add the method as a default method on the trait, which keeps things simple for the implementer.

I think that mandates a separate PR :)

Reply:
> I think this is equiv to the other topic, the overseer processes the messages in sequence as part of a `mpsc::channel()`...

That's a very good point. I'd forgotten that we would be blocking on the overseer in that case instead of directly on the peers. That may be something to consider changing in the future, but for now, feel free to ignore these suggestions; as you say, concurrency doesn't get us much at the moment.